### PR TITLE
WIP: Add FewShot experiment pipeline for BirdSet datasets

### PR DESCRIPTION
**Pull request by raphaelschwinger**
We would like to train a linear network on top of the embeddings generated from a "FoundationModel"  in a FewShot setting. From either the train or directly from the soundscape test set **k** samples per class should be selected and used to train the classification head of the model. The rest of the model stays frozen.

### Train on focal
- [x] merged #17 into this branch, review it first!
- [x] select k samples per class from train set
- [x] train classifier till convergence
- [x] eval on test set
- [x] enable validation
- [x] play with hyperparameter to find best config for BEATs
- [ ] train with augmentations
- [ ] extent experiment to all datasets
- [ ] extent to all models

---

### Train on soundscape directly
- [ ] select k samples per class from test set (maybe use `test` and not `test5s` to sample them)
- [ ] train classifier till convergence
- [ ] eval on (remaining) test set
- [ ] think about a unified validation split?
- [ ] extend experiments to other models
- [ ] extend experiments to other datasets

### CODE REVIEW

**Comment by The-Mats** [comment]
If I understand it correctly:
- We want to use XC (Probably XCL) to train the linear classifier. Like in Birdset this is done separately and the checkpoint is then used for the testing phase of the Soundscapes such as HSN. When looking at the BirdSet experiments both phases seem to be done in multilabel (One-Hot-Encoding during training). To make it into a FS experiment I would guess that the XCL should only provide 32 samples per class then. 
- For multiclass I don't quiet understand how we want to do this (training is trivial)
- Training FS on test data directly corresponds to Dedicated Training (DT) in the Birdset paper right (except only 32 samples p. c.)?

I have and will continue to look into how to realize the first point while using the embedding modules!

**Comment by raphaelschwinger** [comment]
Hi @XgamerTV 

> - We want to use XC (Probably XCL) to train the linear classifier. Like in Birdset this is done separately and the checkpoint is then used for the testing phase of the Soundscapes such as HSN. When looking at the BirdSet experiments both phases seem to be done in multilabel (One-Hot-Encoding during training). To make it into a FS experiment I would guess that the XCL should only provide 32 samples per class then. 

Each BirdSet soundscape dataset (e.g. HSN) contains also a 'train' split containing a subset from XCL with just the species present that are in the soundscape data. This split is what I meant with "train on XC".
> 
> - For multiclass I don't quiet understand how we want to do this (training is trivial)

BirdSet also provides fully annotated soundscape data, from this we could make a multiclass classification problem, but I guess its better to concentrate on multilabel with the `test_5s` split.
> 
> - Training FS on test data directly corresponds to Dedicated Training (DT) in the Birdset paper right (except only 32 samples p. c.)?
> 

DT is training from this XCL subset for only the occurring species. So still XC data for training, soundscape for testing.


Sorry the issue is not well written. I suggest the following approach: We concentrate on multilabel classification for now. Let's do the following experiments:
1. Linear probing: Train the linear classifier like the DT in BirdSet, so XC subset, eval on soundscape
2. FineTune: like DT in BirdSet
3. FewShot training a linear classifier on DT
4. FewShot training a linear classifier on soundscape data directly.

Make a new PR for each of the first two experiments. The later two stay in this PR. Does this make sense?

**Comment by The-Mats** [comment]
Yeah this makes sense! Thanks for the response, I just saw it as I thought I wouldn't get one today and already created some configs for pretraining on XCL but as far as I can tell we don't need them as we use the train split. I have had this weird error where smart sampling doesnt finish and the program just freezes when using XCL (And LT experiments) but couldn't fix it with the debugger. I will still push my stuff (we can remove it if we don't need it) and start a new PR as the freezing doesn't occur for DT 👍

**Comment by raphaelschwinger** [comment]
I updated the issue description.
